### DISCUSSION TITLE
2006/07/14/%E4%B8%80%E4%B8%AA-ivi-%E9%A9%B1%E5%8A%A8%E7%A8%8B%E5%BA%8F%E7%9A%84%E5%BC%80%E5%8F%91%E8%BF%87%E7%A8%8B

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#91] [Announcements]
# 2006/07/14/%E4%B8%80%E4%B8%AA-ivi-%E9%A9%B1%E5%8A%A8%E7%A8%8B%E5%BA%8F%E7%9A%84%E5%BC%80%E5%8F%91%E8%BF%87%E7%A8%8B

这篇文章是我本科的毕业论文，它只是对我实习工作的记录，描述了开发一个IVI驱动程序的全过程，谈不上什么技术含量。

https://blog.qizhen.xyz/2006/07/14/%E4%B8%80%E4%B8%AA-ivi-%E9%A9%B1%E5%8A%A8%E7%A8%8B%E5%BA%8F%E7%9A%84%E5%BC%80%E5%8F%91%E8%BF%87%E7%A8%8B

**Comment by ruanqizhen** [comment]
之前博客下的留言：

kk295kk
2012年11月20日 11:23 编辑
请问IVI instrument driver wizard创建驱动程序只限于GPIB\SERIAL\VXI这几种接口吗？如果要开发PXI接口的怎么办呢？

回复
Lu
2006年08月12日 09:26 编辑
QQ:50335337,联系我
想工作

回复

**Comment by ruanqizhen** [comment]
> **问题**：  
>     “阮先生,你好,我看过你写的关于IVI的文章,写得很详细,有个问题想请教一下,我现在是用LabVIEW的,CVI没有用过,我应该也可以用LabVIEW来开发自己的IVI仪器驱动吧?目前我已经安装了IVI Compliance Package (ICP) 2.5,但是如果要使用自己开发的IVI驱动能出现在MAX上面,是否还需要一些Toolset?比如IVI Installer Creator and the IVI Specific Driver Test Suite之类的?还是我可以用其他什么方法做到.好像那些Toolset不是免费的,我有没有方法可以尽量不用购买什么工具包就可以使用自己的IVI仪器驱动呢?谢谢!  
>     阮先生你好,我是问你IVI问题的那个人,我想到一个方法不知道能不能行得通,如果说我只是使用IVI驱动里某一类驱动程序(比如信号源)很简单的通用属性(比如只是使用到信号源的二个通用属性:频率和功率,全部信号源都支持这二个属性的设置),那我可不可以在NI的网站上下载某一型号信号源的IVI驱动(我手上有另外一种不同型号的信号源,但没有发现我手头上这个信号源的IVI驱动程序),然后安装这个IVI驱动,把其他特有的属性去掉,只留下共用的属性,然后用这个下载的IVI驱动来驱动我手头上这个信号源(假设二种信号源在设置频率和功率时所使用到的命令都是完全相同的,尽管它们的型号是不一样的)?不知道有没有把问题说清楚,还请多多指教.另外,不知道在这里问问题合不合适.谢谢”

**回答：**  
    目前不能用 LabVIEW 编写 IVI 仪器驱动程序，只能用 C 语言编写 IVI。LabVIEW IVI 驱动程序是在已有的 C IVI 驱动程序基础上包装一下得来的，不能在没有 C IVI 的时候直接去开发 LabVIEW IVI。  
    现存的直接使用 LabVIEW 编写仪器驱动程序大多是符合 VXIpnp 标准的仪器驱动程序。如果你用不到互换性，VXIpnp 仪器驱动程序就可以满足要求了。目前，IVI 的可互换性的应用也不是很多，也许你可以考虑直接编写 LabVIEW VXIpnp 仪器驱动程序。

    如果要开发 IVI 驱动程序，最好要有最新版的 CVI 和 IVI Compliance Package (ICP)。其它工具包都可以不用。

    IVI Spacific Driver 是针对某一型号的仪器专用的 IVI 驱动程序。它是不能被用到其它型号的一起上的，就算是只用到通用属性也不行。原因有两个，一个是每个仪器的命令可能是不同的，另一个是驱动程序在初始化的时候就会检查仪器是不是自己所支持的。如果两个型号的仪器所用的命令完全相同，你也好要改一下 Init 函数的源代码才能把驱动用到另一台仪器上。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

肖杰
2011年11月22日 16:35 编辑
请教阮老师，我是电子科大在校学生，请问有没有相关仪器互换的资料呢？我现在的方向是做仪器互换，根据IVI标准，应该就是做仪器类驱动，完全不知道如何下手，不知道您是否能提点一二？或者发点相关资料给我也行，我的邮箱jacksignal@sina.com，谢谢阮老师

回复
ruanqizhen
2011年11月24日 16:09 编辑
我有些年没有做这方面的工作了，我了解的信息也许过时了：IVI方面的资料非常少，因为做的人不多。我的bolg上可以搜到一些很陈旧的内容。估计最权威的还是参考 IVI Foundation 网站上提供的文档。他们的文档还是比较详细的。足够了解IVI开发的全过程了。

回复
Pingback： 一个 IVI 驱动程序的开发过程 « 凡人琐忆 编辑
Qizhen
2009年02月25日 09:48 编辑
你可以首先询问这家仪器公司，问他们是否提供相应的驱动程序。然后再查看一下网页 http://www.ni.com/idnet 。如果都没有，大概就是没有现成的驱动程序了。如果是在LabVIEW下使用，可以就用LabVIEW开发驱动程序。基本上调用VISA相关的函数就可以编写出驱动程序了，可以参考一下其它的驱动程序来编写。

回复
弦
2009年02月24日 17:09 编辑
请教阮先生: 如何通过Lan来实现Labview控制网络分析仪N5230A?有没有现成的驱动可供下载?从哪里可以下载?找了很多地方都下不到.如果没有现成的如何自己编写?用什么语言编写? 谢谢您!!

回复
Unknown
2007年01月20日 15:57 编辑
阮先生，谢谢您的回答，感觉稍微有点思路了。谢谢。
BTW，这个网站好难打开…

回复

**Comment by ruanqizhen** [comment]
问题(Martin)：

现在想请教有关IVI的一个问题: 我在一个系统中要用到AC SOURCE并且需要使用IVI的可互换性,但是IVI并没有发布AC power 的class driver,这可能需要自己开发AC power class driver. 但我却看到NI或其他仪器公司的网站上提供基于IVI的specific driver(当然是AC source的),那labview在调用这些ac power的 specific driver时是否是通过class driver呢? 我认为不是的,那这样也应该不具有IVI的可互换性,请问是这样吗?

**回答：**    

IVI 是还没有 AC Power 的 Class Drver，但是还是可以开发 AC Power 的 Specific Drver 的。只不过这样的 Specific Drver 就不具备可互换性了。如果一定能够需要可互换性，确实需要自己开发 Class Drver。以前好像有厂商提过要自己开发 Class Drver，但是我后来就没收到过更新的消息。自己开发 Class Drver 没有很多可以借助的工具，也没有开放的代码可供参考，恐怕困难会大一些。

**Reply by ruanqizhen** [reply]
之前博客下的留言：

方
2011年04月4日 13:20 编辑
您好，今天刚收到您的书，目前只看的了序和前言，感觉不错 。我是一个学生，目前在做一个项目，测试相关的参数，发现LabVIEW是入门易精通难的，希望能与您有交流的机会！！！

回复
ruanqizhen
2011年04月4日 18:31 编辑
多多交流

回复
Tom Lu
2007年04月8日 14:33 编辑
以前接触过一个程序采用Plug-in的方式来处理这个问题的，就是两种仪器的控制用不同的vi但接口是一样的，可以通过配置决定调用那一个仪器的vi。

回复
Unknown
2007年04月3日 17:01 编辑
感谢回复以及大家的谈论.高杰的办法也是一个不错的方法.
我现在打算用VI server的方法来实现一些仪器的互换,也就阮回复的那样.
工作还在进行中.
另外祝新婚快乐.

我还看了你写的关于 荨麻疹 经历, 感触很深,因为今年过年时我的妈妈也得了这种病,但当时对它不了解,以为就是
一般的皮肤过敏,吃点药就好了.后来才得知这种病要是变成慢性的会很折麽人的. 所以现在很担心,也后悔为什么没
在家多陪陪老妈.

健康.

回复
Unknown
2007年04月3日 17:01 编辑
感谢回复以及大家的谈论.高杰的办法也是一个不错的方法.
我现在打算用VI server的方法来实现一些仪器的互换,也就阮回复的那样.
工作还在进行中.
另外祝新婚快乐.

我还看了你写的关于 荨麻疹 经历, 感触很深,因为今年过年时我的妈妈也得了这种病,但当时对它不了解,以为就是
一般的皮肤过敏,吃点药就好了.后来才得知这种病要是变成慢性的会很折麽人的. 所以现在很担心,也后悔为什么没
在家多陪陪老妈.

健康.

回复
Qizhen
2007年04月2日 16:21 编辑
回 怒放的生命关于具备什么样的条件才能进一个公司这很难说的。每个面试官可能都有不同的观点，公司之间大概差距更大。
现在LabVIEW的应用也不是很普及，我们公司招来的新员工大多都是没有用过LabVIEW的。所以是否会LabVIEW可能不想你想得那么重要。
我个人觉得，当前的学习和工作成绩可能对将来的求职影响最大吧。比如说，如果你现在成绩总是在系里前几名，实践能力也不差，心目中有五六家比较理想的公司，都去面试一下，肯定会被其中的某系公司看中的。能把现在的事情做得很优秀，就不用太担心以后的工作了。

回复
Qizhen
2007年04月2日 15:55 编辑
自己开发 IVI Class Driver 不但麻烦，而且其他公司也不见得认同你的标准。开发出来的 Class Driver 只能配合自己写的 Specfic Driver。
所以我觉得高杰的办法也不错，比开发新的 Class Driver 效率高。另外我以前还用过一个方法，适合简单的一起操作。如果一个系统之用到很简单的几个一起操作，比如：“打开”、“配置”、“读数据”“关闭”。可以针对你用到的所以得仪器都编写四个这样的函数（VI），保持参数的个数和类型一致。在应用程序中根据仪器的不同动态的调用相应的每个仪器的函数。这种方法也让系统具备了可互换性，并且实现起来也相对简单。

回复
清扬
2007年04月2日 12:46 编辑
阮先生,我想问下,如果想进NI的话,除了精通labview之外,还需要懂什么呢?我现在还在上研究生,非常喜欢labview,希望能进一个象NI这样的大公司,能和您这样的真正热爱技术的人共事,希望您能给点指导.现在在学校很迷茫,虽然还有一年就毕业了,可还不知道自己究竟需要具备什么样的素质才能被理想中的公司录用.我曾经看过您的一篇关于招聘的文章,我想前辈的经验一定会更有利于我们职业规划的.谢谢~

回复
杰
2007年03月29日 22:55 编辑
自己开发标准而全面的某一种类的IVI Class Driver可能真的麻烦了一点。甚至于有的时候，某些仪器不支持IVI。不过其实在真正的工程应用中，还是可以方便的实现的。
所谓IVI就是具有相同功能的仪器在系统中的硬件互换而不影响软件。（我自己的理解，可能有些许偏差。呵呵。）
那么，很简单的做一个小的ACESS甚至于一个TXT文件把需要用到的Program command分类的全部写进去。程序初始化的时候用IDN识别硬件，作为索引，在软件中需要用到的地方找出相应的command发给相应的硬件即可。
因为不同的硬件厂商虽然对仪器的操作命令不同，但IDN？这样的通用命令是都支持的。
我开发的一套ATS里面就是这么干的，呵呵。里面大概能用3种不同的示波器、5种不同的DC POWER，2种不同的网络分析仪。其实真正在工程中应用之后，都做成小模块了，会发现即使再多种类的不同厂家的相同类的仪表，要把软件做成通用的，还是很方便的。看一下仪器的manual，在记录command的文件里面添加一行数据即可。不过我那个是用VB写的，LV我还是新手上路。但我觉得这种编程的思想还是通用。
可能俺的这种方法太老土，不过还是有效。见笑了。

回复

**Comment by ruanqizhen** [comment]
# 可互换虚拟仪器驱动程序的开发

## 一. 仪器驱动程序的发展

    仪器驱动程序是连接仪器硬件与高层应用程序的桥梁。使用仪器驱动程序可以把高层应用函数从仪器接口函数中分离出来，从而简化了高层应用软件的开发难度，提高了应用软件的可维护性。  
    仪器驱动程序的发展大致经历了以下几个阶段：  
    1. 早期仪器驱动程序 没有任何统一的规范，功能较为简单、直接。由于不同的仪器驱动程序的实现方法和接口函数大相径庭，用户往往觉得它们难以琢磨，结果使得它们很难被应用和修改。用户需要的是容易使用，便于修改，并且遵循相同标准的驱动程序。只有这样，他们才有可能把不同厂商的仪器集成在同一系统中使用。为了实现这一目标，针对仪器驱动程序制定的标准陆续出现，比如下文提到的这几种比较重要的规范。  
    2. IEEE 488.2规范 定义了用于仪器的辨认和自检等的命令，但是仅有这几种命令还远远不能解决仪器驱动程序编写繁复、阅读困难等缺点。  
    3. SCPI（Standard Commands for Programmable Instruments）规范 定义了通用的仪器命令结构，对命令按功能进行了归类分组。目前，绝大多数新仪器都是基于字符串命令的仪器，并且采用了SCPI规范。遗憾的是，虽然他们都使用了SCPI的结构，但具体到每个功能，不同厂商仍然使用各不相同的命令字符串。  
    4. VXI总线即插即用（VXIplug&play）规范（简称VPP规范）由VPP系统联盟制定。VPP规范侧重于仪器的“互操作性（interoperability）”，即用户可以方便地把不同厂商的仪器集成到自己系统中。为此，VPP规范详细定义了仪器驱动程序的内部设计模型与外部接口模型，规定了虚拟仪器驱动程序的设计要求及与系统中其他程序模块之间的相互关系。但是，VPP规范仍然没能解决工程师建造测试系统时遇到的困扰：一是软件的开发和维护费用过于昂贵，二是难以快速升级已有的系统。解决这样的问题需要仪器驱动程序具有“互换性”。  
    5. IVI（Interchangeable Virtual Instrument）驱动程序规范 扩展了VPP仪器驱动程序的标准，并增加了仪器的可互换性、仿真和状态缓存等特点。IVI驱动程序规范由IVI基金会（Interchangeable Virtual Instrument Foundation）制定。它是由最终用户、系统集成商和仪器制造商共同组成的一个开放联盟，几乎所有的大型仪器生产公司（诸如：National Instruments，Agilent，Tektronix）都已经成为其成员。其职能是为测试仪器制定编程规范，从而实现不同型号仪器之间的互换，并且提高系统性能、降低开发成本。

## 二. IVI系统结构

    为了实现互换性，IVI基金会将同类仪器的共性提取出，并作了规范。目前已经发布的八类仪器规范是：示波器（IviScope）、数字万用表（IviDmm）、信号发生器（IviFGen）、直流电源（IviDCPower）、开关矩阵/多路复用器（IviSwitch）、功率表（IviPwrMeter）、频谱分析仪（IviSpecAn）和射频信号发生器（IviRFSigGen），其他类型仪器的规范也将被陆续制定发布。每一类的仪器都有各自的类驱动程序（IVI Class Driver）。类驱动程序包含了该类仪器通用的各种属性和操作函数。运行时，类驱动程序通过调用每台仪器的专用驱动程序（IVI Specific Driver）中相应的函数来控制仪器。


图1： IVI体系结构

  
    图1是IVI体系结构框图：  
    应用程序可以直接调用专用驱动程序来控制仪器。但是为了实现仪器互换，应用程序应该首先调用类驱动程序，类驱动程序检查IVI配置文件以确定应该使用的专用驱动程序。若系统中的仪器被更换，只需适当修改IVI配置文件，而应用程序无需任何改动，因而实现了测试系统的通用性。

##   
三. IVI驱动程序的特点和优势

    1. 互换性。IVI驱动程序的互换性至少为我们带来了以下几大好处：  
        a) 易于使用。所用的IVI驱动程序都使用通用的接口，易于理解，也就不再要求应用程序的开发人员必须掌握某一特定仪器的编程方法，从而系统开发获得了更大的硬件独立性。  
b) 降低了系统的维护和升级费用。IVI构架系统可以适用不同的仪器，当仪器陈旧或者有了升级的、高性能或低造价的仪器时，可以任意更换，而不需要改变应用程序。  
        c) 代码共享。IVI构架允许部门和设备之间方便地复用及共享测试代码，并且不需使用相同型号的仪器硬件。  
    2. 模拟功能。每个仪器专用驱动程序都具有专门针对本型号仪器的模拟功能。这些模拟功能使得工程师在缺少真实仪器的情况下，可以使用IVI驱动程序的模拟功能来开发、调试应用程序。还可以使用美国国家仪器公司（National Instruments，简称NI）提供的类模拟驱动程序以获得更强大的模拟功能。  
    3. 状态缓存功能。IVI驱动程序可以保存仪器每一属性设置的当前状态。当应用程序试图发送一些冗余命令到仪器时（例如，将仪器的某一属性重新设置为当前值，这些命令显然不会让仪器产生任何变化或动作），IVI驱动程序会跳过这些命令。当前的测试系统中，影响软件执行速度的瓶颈通常在于仪器与计算机接口总线的传输速率，IVI驱动程序的此项功能大大减少了仪器与计算机之间的通讯，从而提升了系统性能。  
    4. 源码开放。高级用户可以直接修改IVI驱动程序的源代码，以对其进行优化或添加功能。  
    5. 免费得到大量的驱动程序。除了生产厂商自行开发的IVI驱动程序，NI公司也为各类常用仪器开发了大量IVI驱动程序，这些程序都可以从NI的网站上免费下载。 此外，NI还提供了用于开发驱动程序的工具包，以简化用户的IVI驱动程序的开发过程。

## 四. IVI驱动程序的设计和开发

    按照提供给用户的接口形式的不同，IVI驱动程序分为：C语言驱动程序， LabVIEW驱动程序，COM驱动程序。虽然同一仪器可以存在接口形式不同的驱动程序，但它们的功能完全相同，用户可以选择使用自己最擅长的形式。目前最普遍使用的仍然是C语言驱动程序；LabVIEW驱动程序是在C语言驱动程序生成的DLL动态库的基础上生成的； 而COM驱动程序则需要重新编写代码。  
    IVI驱动程序规范已经详细制定了驱动程序的构架和接口，因此开发人员不必另行设计程序结构，这避免了驱动程序结构不合理或用户使用不便等问题。但是，开发人员要特别注意和精心设计仪器的功能、控制命令与IVI规范不兼容的部分。  
    使用LabWindows/CVI作为C语言驱动程序的开发工具具有独特的优越条件，因为CVI集成了许多驱动程序开发工具，比如：IVI驱动程序生成向导、属性编辑器等。另外还可以选用NI公司的IDT工具包作为补充。

    下面以较为常用的Fluke 45 数字万用表为例，简单介绍一下开发IVI仪器驱动程序的具体步骤：  
    1． 熟悉仪器与IVI规范。在开发仪器的IVI驱动程序以前，首先要熟悉仪器的有关命令，然后根据其功能进行分类。并且将仪器命令与IVI规范制定的属性、函数对应起来。  
    2． 使用CVI的“仪器驱动程序开发向导”生成驱动程序的框架。开发向导生成的文件一共有4个，其中，“fl45.c”和“fl45.h”分别是C语言源文件和头文件；“fl45.fp”记录函数面板信息；“fl45.sub”记录仪器“属性（attribute）”的信息。  
    3． 删除仪器不支持的扩展功能组。可能会出现仪器不支持某些IVI规范定义的功能组的情况，比如说fl45不具备温度测量功能，这时应该把所有与温度测量有关的函数和属性移出。  
    4． 完善代码。实现IVI规范中仪器可以支持的功能组。“仪器驱动程序开发向导”已经生成了大部分属性和函数的代码。用户需要做的只是到属性的回调函数（Attribute Callback Function）中修改相应的仪器命令。由于每台仪器的命令体系都有其特殊性，用户也不能够完全依赖“开发向导”生成的程序结构，函数和属性的实现需要参照仪器的具体命令体系来实现。  
    5． 编写支持仪器特殊功能的代码。“仪器驱动程序开发向导”不能够生成支持IVI规范定义之外的仪器功能。比如Fl45有对测量值进行修正的功能，IVI规范没有相应的定义，实现这部分功能需要自己设计和编写相应的属性及函数。  
    6． 测试。除了要确保IVI驱动程序函数功能正确，还要验证它是否完全符合IVI规范。IVI驱动程序的开发人员也可以请求NI公司帮助完成程序的最终测试。  
    7． 编写LabVIEW驱动程序。如果已经有了C语言IVI驱动程序，只需要使用NI公司在其网站上提供的 [LabVIEW Interface Generator for LabWindows™/CVI™ Instrument Drivers](http://sine.ni.com/nilex/DisplayLinkAction.do?id=101NILEX) 工具即可方便地生成相应的LabVIEW仪器驱动程序。

    IVI类驱动程序的开发。如果用户希望使用IVI规范定义的八类仪器之外的仪器，并且还要实现通用性（可互换性），那么则需要客户自己开发类驱动程序。类驱动程序的开发与仪器驱动程序的开发过程完全不同，由于用户很少遇到类驱动程序的开发，本文不作详细介绍。

## 五. IVI驱动程序的应用实例

    IVI驱动程序相对于传统仪器驱动程序而言，比较适合于高端应用和大型测试系统的搭建。下面是两个IVI体系的典型应用。

    1． 2002年波音公司在其新设计的测试系统中应用了IVI技术。由于对测试稳定性的要求，波音公司的测试系统一经投入使用，往往要求连续工作几十年而且不能做大的改动。在这么长的使用期内，许多最初设计时使用的仪器都可能停产。一旦仪器损坏，根本无法找到同型号的仪器进行替换，只能使用类似的仪器进行替代。如果使用传统仪器驱动程序，每次更换仪器时，测试软件也必须做相应更新。这样一来不但需要花费大量时间在测试软件的更新上，更重要的是对软件结构的改动很可能会破坏它的一致性。  
    在使用IVI体系结构后，每次替换仪器，只需要在IVI配置文件中指明新的驱动程序即可，测试软件不需要作任何改动。这样既节约了时间，又不会对测试结果产生任何影响。

    2． NI公司在其2005年推出的校准系统中也使用了IVI体系结构。该校准系统将提供给世界各地校准实验室使用，因此，该校准系统必须能供控制型号千差万别的各类标准仪器。IVI体系结构保证了该系统在不改变代码的前提下适应任何实验室的仪器设备。
